File: DevExpress-Demo/CS/GridDemo.Wpf/ModuleResources/StockViewModelBase.cs
using DevExpress.Mvvm.DataAnnotations;
using DevExpress.Xpf.Core;
using DevExpress.Xpf.DemoBase.Helpers;
using System;
using System.Collections.Generic;
using System.Reflection;
using System.Windows.Threading;
using System.Xml.Serialization;

namespace GridDemo {
    [POCOViewModel]
    public abstract class StockViewModelBase {
        const int HistoryLength = 20;
        const int UpdateInterval = 1000;

        protected readonly Random Random = new Random();

        DispatcherTimer timer;

        public StockViewModelBase() {
            ObservableCollectionCore<StockItem> data = StockItems.GetData();
            foreach(StockItem stockItem in data) {
                for(int i = 0; i < HistoryLength; i++)
                    UpdatePrice(stockItem, true);
            }
            Data = data;
        }

        public ObservableCollectionCore<StockItem> Data { get; set; }

        public void StartUpdate() {
            timer = new DispatcherTimer() { Interval = TimeSpan.FromMilliseconds(UpdateInterval) };
            timer.Tick += timer_Tick;
            timer.Start();
        }
        public void StopUpdate() {
            timer.Stop();
            timer.Tick -= timer_Tick;
            timer = null;
        }

        void timer_Tick(object sender, EventArgs e) {
            Data.BeginUpdate();
            foreach(StockItem stockItem in Data)
                UpdatePrice(stockItem, false);
            Data.EndUpdate();
        }
        void UpdatePrice(StockItem stockItem, bool dataGeneration) {
            double newDelta = GetDeltaPrice(stockItem.Price, dataGeneration);
            if (stockItem.Price + newDelta <= 0)
                newDelta = 0;
            if(Math.Sign(stockItem.DeltaPrice) == Math.Sign(newDelta))
                stockItem.DeltaChange = 0;
            else
                stockItem.DeltaChange = Math.Sign(newDelta);
            stockItem.DeltaPrice = newDelta;
            stockItem.Price += stockItem.DeltaPrice;
            if(stockItem.Price < stockItem.LowPrice)
                stockItem.LowPrice = stockItem.Price;
            if(stockItem.Price > stockItem.HighPrice)
                stockItem.HighPrice = stockItem.Price;
            stockItem.DeltaPricePercent = stockItem.DeltaPrice / stockItem.Price;
            stockItem.Volume += Convert.ToInt32(Random.NextDouble() * stockItem.Volume * 0.005 - 0.0025);
            UpdatePriceHistory(stockItem);
        }
        void UpdatePriceHistory(StockItem stockItem) {
            List<double> newPriceHistory = new List<double>(new double[HistoryLength]);
            for(int i = 1; i < HistoryLength; i++)
                newPriceHistory[i - 1] = i < stockItem.PriceHistory.Count ? stockItem.PriceHistory[i] : 0;
            newPriceHistory[HistoryLength - 1] = stockItem.Price;
            stockItem.PriceHistory = newPriceHistory;
        }

        protected abstract double GetDeltaPrice(double currentPrice, bool dataGeneration);
    }

    [XmlRoot("StockItems")]
    public class StockItems : ObservableCollectionCore<StockItem> {
        public static ObservableCollectionCore<StockItem> GetData() {
            XmlSerializer s = new XmlSerializer(typeof(StockItems));
            Assembly assembly = typeof(StockItems).Assembly;
            return (ObservableCollectionCore<StockItem>)s.Deserialize(assembly.GetManifestResourceStream(DemoHelper.GetPath("GridDemo.Data.", assembly) + "StockSource.xml"));
        }
    }

    public class StockItem {
        public string CompanyName { get; set; }
        public double Price { get; set; }
        public int Volume { get; set; }
        public double LowPrice { get; set; }
        public double HighPrice { get; set; }
        public double DeltaPrice { get; set; }
        public double DeltaPricePercent { get; set; }
        public int DeltaChange { get; set; }

        public StockItem() {
            PriceHistory = new List<double>();
        }
        public List<double> PriceHistory { get; set; }
    }
}

File: DevExpress-Demo/VB/GridDemo.Wpf/ModuleResources/StockViewModelBase.vb
Imports DevExpress.Mvvm.DataAnnotations
Imports DevExpress.Xpf.Core
Imports DevExpress.Xpf.DemoBase.Helpers
Imports System
Imports System.Collections.Generic
Imports System.Reflection
Imports System.Windows.Threading
Imports System.Xml.Serialization

Namespace GridDemo

    <POCOViewModel>
    Public MustInherit Class StockViewModelBase

        Const HistoryLength As Integer = 20

        Const UpdateInterval As Integer = 1000

        Protected ReadOnly Random As Random = New Random()

        Private timer As DispatcherTimer

        Public Sub New()
            Dim data As ObservableCollectionCore(Of StockItem) = StockItems.GetData()
            For Each stockItem As StockItem In data
                For i As Integer = 0 To HistoryLength - 1
                    UpdatePrice(stockItem, True)
                Next
            Next

            Me.Data = data
        End Sub

        Public Property Data As ObservableCollectionCore(Of StockItem)

        Public Sub StartUpdate()
            timer = New DispatcherTimer() With {.Interval = TimeSpan.FromMilliseconds(UpdateInterval)}
            AddHandler timer.Tick, AddressOf timer_Tick
            timer.Start()
        End Sub

        Public Sub StopUpdate()
            timer.Stop()
            RemoveHandler timer.Tick, AddressOf timer_Tick
            timer = Nothing
        End Sub

        Private Sub timer_Tick(ByVal sender As Object, ByVal e As EventArgs)
            Data.BeginUpdate()
            For Each stockItem As StockItem In Data
                UpdatePrice(stockItem, False)
            Next

            Data.EndUpdate()
        End Sub

        Private Sub UpdatePrice(ByVal stockItem As StockItem, ByVal dataGeneration As Boolean)
            Dim newDelta As Double = GetDeltaPrice(stockItem.Price, dataGeneration)
            If stockItem.Price + newDelta <= 0 Then newDelta = 0
            If Math.Sign(stockItem.DeltaPrice) = Math.Sign(newDelta) Then
                stockItem.DeltaChange = 0
            Else
                stockItem.DeltaChange = Math.Sign(newDelta)
            End If

            stockItem.DeltaPrice = newDelta
            stockItem.Price += stockItem.DeltaPrice
            If stockItem.Price < stockItem.LowPrice Then stockItem.LowPrice = stockItem.Price
            If stockItem.Price > stockItem.HighPrice Then stockItem.HighPrice = stockItem.Price
            stockItem.DeltaPricePercent = stockItem.DeltaPrice / stockItem.Price
            stockItem.Volume += Convert.ToInt32(Random.NextDouble() * stockItem.Volume * 0.005 - 0.0025)
            UpdatePriceHistory(stockItem)
        End Sub

        Private Sub UpdatePriceHistory(ByVal stockItem As StockItem)
            Dim newPriceHistory As List(Of Double) = New List(Of Double)(New Double(19) {})
            For i As Integer = 1 To HistoryLength - 1
                newPriceHistory(i - 1) = If(i < stockItem.PriceHistory.Count, stockItem.PriceHistory(i), 0)
            Next

            newPriceHistory(HistoryLength - 1) = stockItem.Price
            stockItem.PriceHistory = newPriceHistory
        End Sub

        Protected MustOverride Function GetDeltaPrice(ByVal currentPrice As Double, ByVal dataGeneration As Boolean) As Double
    End Class

    <XmlRoot("StockItems")>
    Public Class StockItems
        Inherits ObservableCollectionCore(Of StockItem)

        Public Shared Function GetData() As ObservableCollectionCore(Of StockItem)
            Dim s As XmlSerializer = New XmlSerializer(GetType(StockItems))
            Dim assembly As Assembly = GetType(StockItems).Assembly
            Return CType(s.Deserialize(assembly.GetManifestResourceStream(DemoHelper.GetPath("GridDemo.Data.", assembly) & "StockSource.xml")), ObservableCollectionCore(Of StockItem))
        End Function
    End Class

    Public Class StockItem

        Public Property CompanyName As String

        Public Property Price As Double

        Public Property Volume As Integer

        Public Property LowPrice As Double

        Public Property HighPrice As Double

        Public Property DeltaPrice As Double

        Public Property DeltaPricePercent As Double

        Public Property DeltaChange As Integer

        Public Sub New()
            PriceHistory = New List(Of Double)()
        End Sub

        Public Property PriceHistory As List(Of Double)
    End Class
End Namespace
